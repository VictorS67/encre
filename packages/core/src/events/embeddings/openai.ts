import {
  OpenAI as OpenAIClient,
  ClientOptions as OpenAIClientOptions,
} from 'openai';
import type { RequestOptions as OpenAIClientRequestOptions } from 'openai/core';
import { type SecretFields, type SerializedFields } from '../../load/keymap.js';
import { getEnvironmentVariables } from '../../utils/environment.js';
import { type EmbedResult } from '../output/provide/index.js';
import {
  BaseEmbeddings,
  type BaseEmbeddingsCallOptions,
  type BaseEmbeddingsParams,
} from './base.js';

/**
 * Parameters for the OpenAI embeddings, including model and API key specifics.
 */
export interface OpenAIEmbeddingsParams extends BaseEmbeddingsParams {
  /**
   * ID of the model to use. You can use the [List models]({@link https://platform.openai.com/docs/api-reference/models/list}}
   * API to see all of your avaliable models, or see our
   * [Model overview]({@link https://platform.openai.com/docs/models/overview}) for description of them.
   */
  modelName: string;

  /**
   * The number of dimensions the resulting output embeddings should have.
   * Only supported in `text-embedding-3` and later models.
   */
  dimensions?: number;

  /**
   * A unique identifier representing your end-user, which can help OpenAI to monitor
   * and detect abuse.
   *
   * [Learn more]({@link https://platform.openai.com/docs/guides/safety-best-practices})
   */
  user?: string;

  /**
   * API key to use when making requests to OpenAI. Defaults to the value of
   * `OPENAI_API_KEY` environment variable.
   */
  openAIApiKey?: string;

  /**
   * Timeout for the API requests.
   */
  timeout?: number;
}

/**
 * Call options for OpenAI embeddings, including format and additional request options.
 */
export interface OpenAIEmbeddingsCallOptions extends BaseEmbeddingsCallOptions {
  /**
   * Format of the encoding for the embeddings, either as 'float' or 'base64'.
   */
  encondingFormat?: 'float' | 'base64';

  /**
   * Additional options to pass to the underlying axios request.
   */
  options?: OpenAIClientRequestOptions;
}

/**
 * Class to handle embeddings via the OpenAI API, extending the base embeddings functionality.
 *
 * @example
 * ```typescript
 * const embeddings = new OpenAIEmbeddings({ modelName: 'text-embedding-ada-002' });
 * const options = { encondingFormat: 'float' };
 * const response = await embeddings.invoke('Hello!', options);
 * console.log(response);
 * ```
 */
export class OpenAIEmbeddings<
    CallOptions extends
      OpenAIEmbeddingsCallOptions = OpenAIEmbeddingsCallOptions,
  >
  extends BaseEmbeddings<CallOptions>
  implements OpenAIEmbeddingsParams
{
  _isSerializable = true;

  get _secrets(): SecretFields | undefined {
    return {
      openAIApiKey: 'OPENAI_API_KEY',
      organization: 'OPENAI_ORGANIZATION',
    };
  }

  get _aliases(): SerializedFields | undefined {
    return {
      modelName: 'model',
      openAIApiKey: 'openai_api_key',
    };
  }

  static _name(): string {
    return 'OpenAI';
  }

  _embeddingsType(): string {
    return 'openai';
  }

  /**
   * ID of the model to use. You can use the [List models]({@link https://platform.openai.com/docs/api-reference/models/list}}
   * API to see all of your avaliable models, or see our
   * [Model overview]({@link https://platform.openai.com/docs/models/overview}) for description of them.
   */
  modelName = 'text-embedding-ada-002';

  /**
   * The number of dimensions the resulting output embeddings should have.
   * Only supported in `text-embedding-3` and later models.
   */
  dimensions?: number | undefined;

  /**
   * A unique identifier representing your end-user, which can help OpenAI to monitor
   * and detect abuse.
   *
   * [Learn more]({@link https://platform.openai.com/docs/guides/safety-best-practices})
   */
  user?: string | undefined;

  /**
   * The max number of concurrent calls that can be made.
   * Defaults to 2.
   */
  maxConcurrency?: number | undefined;

  /**
   * The max number of retries that can be made for a single call.
   */
  maxRetries?: number | undefined;

  /**
   * Timeout for the API requests.
   */
  timeout?: number;

  /**
   * API key to use when making requests to OpenAI. Defaults to the value of
   * `OPENAI_API_KEY` environment variable.
   */
  openAIApiKey?: string;

  /**
   * Identifier for organization sometimes used in API requestsã€‚
   */
  organization?: string;

  /**
   * OpenAI API Client.
   * @internal
   */
  private _client: OpenAIClient;

  /**
   * OpenAI API request options.
   * @internal
   */
  private _clientOptions: OpenAIClientOptions;

  constructor(
    fields?: Partial<OpenAIEmbeddingsParams> & {
      configuration?: OpenAIClientOptions;
    }
  ) {
    fields = {
      modelName: fields?.modelName ?? 'text-embedding-ada-002',
      maxConcurrency: fields?.maxConcurrency ?? 2,
      maxRetries: fields?.maxRetries,
      dimensions: fields?.dimensions,
      user: fields?.user,
      timeout: fields?.timeout,
      ...fields,
    };

    super(fields);

    this.openAIApiKey =
      fields?.openAIApiKey ?? getEnvironmentVariables('OPENAI_API_KEY');

    if (!this.openAIApiKey) {
      throw new Error('OpenAI API Key not found');
    }

    this.modelName = fields?.modelName ?? this.modelName;

    this.maxConcurrency = fields?.maxConcurrency ?? 2;
    this.maxRetries = fields?.maxRetries;
    this.dimensions = fields?.dimensions;
    this.user = fields?.user;
    this.timeout = fields?.timeout;

    this.organization =
      fields?.configuration?.organization ??
      getEnvironmentVariables('OPENAI_ORGANIZATION');

    this._clientOptions = {
      apiKey: this.openAIApiKey,
      organization: this.organization,
      baseURL: fields?.configuration?.baseURL,
      dangerouslyAllowBrowser: true,
      defaultHeaders: fields?.configuration?.defaultHeaders,
      defaultQuery: fields?.configuration?.defaultQuery,
      ...fields?.configuration,
    };
  }

  /**
   * Retrieves the parameters required for creating an embedding with the OpenAI API.
   * This method extracts and prepares the parameters based on the model configurations
   * and any additional options specified in the call.
   *
   * @param options Optional serialized call options which may override default model settings.
   * @returns An object containing the necessary parameters for the OpenAI API request,
   *          excluding the 'input' parameter which is handled separately.
   */
  getParams(
    options?: this['SerializedCallOptions']
  ): Omit<OpenAIClient.EmbeddingCreateParams, 'input'> {
    return {
      model: this.modelName,
      dimensions: this.dimensions,
      encoding_format: options?.encondingFormat,
      user: this.user,
    };
  }

  _identifyParams(): Omit<OpenAIClient.EmbeddingCreateParams, 'input'> & {
    model_name: string;
  } & OpenAIClientOptions {
    return {
      model_name: this.modelName,
      ...this.getParams(),
      ...this._clientOptions,
    };
  }

  /**
   * Embeds a document using the OpenAI API.
   * @param document The document to embed.
   * @param options Serialized call options.
   * @returns A promise resolving to the embedding result.
   * @internal
   * @example
   * ```typescript
   * const embeddings = new OpenAIEmbeddings({ modelName: 'text-embedding-ada-002' });
   * const result = await embeddings._embed("Hello, world!", {});
   * console.log(result.embedding);
   * ```
   */
  async _embed(
    document: string,
    options: this['SerializedCallOptions']
  ): Promise<EmbedResult> {
    const params = this.getParams(options);

    const response = await this.completionWithRetry(
      {
        ...params,
        input: document,
      },
      {
        signal: options?.signal,
        ...options?.options,
      }
    );

    return {
      embedding: response.data[0].embedding,
      embedOutput: {
        tokenUsage: {
          promptTokens: response.usage.prompt_tokens,
          totalTokens: response.usage.total_tokens,
        },
      },
    };
  }

  /**
   * Handles retries for the embedding creation in case of failures.
   * @param request The embedding creation parameters.
   * @param options Optional client request configurations.
   * @returns A promise resolving to the embedding creation response.
   * @example
   * ```typescript
   * const embeddings = new OpenAIEmbeddings({ modelName: 'text-embedding-ada-002' });
   * const requestParams = { input: "Example document", model: "text-embedding-ada-002" };
   * const response = await embeddings.completionWithRetry(requestParams);
   * console.log(response.data[0].embedding);
   * ```
   */
  async completionWithRetry(
    request: OpenAIClient.EmbeddingCreateParams,
    options?: OpenAIClientRequestOptions
  ): Promise<OpenAIClient.Embeddings.CreateEmbeddingResponse> {
    const requestOptions: OpenAIClientRequestOptions =
      this._getRequestOptions(options);
    return this.caller.call(async () => {
      const res = await this._client.embeddings.create(request, requestOptions);
      return res;
    });
  }

  /**
   * Builds request options for the API call, merging default client options with
   * method-specific options.
   * @param options Optional additional request configurations.
   * @returns The combined request options.
   * @internal
   */
  private _getRequestOptions(
    options?: OpenAIClientRequestOptions
  ): OpenAIClientRequestOptions {
    if (!this._client) {
      const params = {
        ...this._clientOptions,
        timeout: this.timeout,
        maxRetries: 0,
      };

      if (!params.baseURL) {
        delete params.baseURL;
      }

      this._client = new OpenAIClient(params);
    }

    const requestOptions: OpenAIClientRequestOptions = {
      ...this._clientOptions,
      ...options,
    };

    return requestOptions;
  }
}
