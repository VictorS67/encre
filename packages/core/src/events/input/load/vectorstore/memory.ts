import { similarity as mlDistanceSimilarity } from 'ml-distance';
import { maximalMarginalRelevance } from '../../../../utils/math.js';
import { getRecordId } from '../../../../utils/nanoid.js';

import { Context } from '../docs/index.js';
import { BaseVectorStore, type BaseVectorStoreField } from './base.js';

/**
 * Represents the fields for the MemoryVectorStore which include a list of vectors and a
 * similarity function definition, with optional filters.
 */
export interface MemoryVectorStoreField extends BaseVectorStoreField {
  /**
   * A list of all vectors currently stored in the memory. Each vector includes its content,
   * embedding, and metadata.
   */
  vectors: Vector[];

  /**
   * The function used to calculate the similarity between two vectors. This is typically a
   * cosine similarity function from a machine learning library.
   */
  similarity: typeof mlDistanceSimilarity.cosine;

  /**
   * Optional filters that can be applied during vector retrieval operations to limit the
   * results based on custom logic.
   */
  filter?: MemoryVectorStoreFilterType;
}

/**
 * Defines the parameters for the delete operation in a MemoryVectorStore. This allows
 * specifying vector IDs to delete or applying a filter to determine which vectors to delete.
 * @template T The type of the filter used in the delete operation.
 */
export interface MemoryDeleteParams<T> {
  /**
   * Optional list of vector IDs to be deleted. If specified, only vectors with these IDs
   * will be considered for deletion.
   */
  ids?: string[];

  /**
   * An optional filter used to determine which vectors should be deleted. The filter must be
   * compatible with the type of filtering logic implemented in the vector store.
   */
  filter?: T;
}

/**
 * Represents a vector within the MemoryVectorStore. Each vector has an identifier, content
 * text, a numerical embedding, and any associated metadata.
 */
export type Vector = {
  /**
   * A unique identifier for the vector.
   */
  id: string;

  /**
   * The content associated with the vector, typically a string of text from which the
   * embedding is derived.
   */
  content: string;

  /**
   * The numerical embedding of the content, represented as an array of numbers. Embeddings
   * are usually generated by a machine learning model to capture semantic meanings of the
   * content.
   */
  embedding: number[];

  /**
   * A dictionary containing metadata associated with the vector. This can include additional
   * information like tags, categories, or any other contextual data that supports the
   * application's use case.
   */
  metadata: Record<string, unknown>;
};

/**
 * Defines the filter type for the MemoryVectorStore. Filters can include a function to
 * determine inclusion based on context, and an option to include embeddings in the returned data.
 */
export type MemoryVectorStoreFilterType = {
  /**
   * An optional function that determines whether a given context meets certain criteria.
   * This function should be asynchronous and return a boolean indicating whether the context
   * passes the filter. This allows for dynamic and complex filtering logic based on the
   * context's properties.
   */
  filterFunc?: (context: Context) => Promise<boolean>;

  /**
   * Specifies whether embeddings should be included in the retrieval results. If set to true,
   * the embeddings associated with the vectors are included in the returned data, allowing
   * further operations or analyses that require access to the raw embeddings.
   */
  includeEmbeddings?: boolean;
};

/**
 * Provides an in-memory implementation of a vector store that supports operations such as adding,
 * deleting, and searching vectors based on similarity or relevance. This class is suitable for
 * lightweight applications or environments where persistence is not required.
 *
 * @example
 * ```typescript
 * // Create an instance of the MemoryVectorStore
 * const vectorStore = new MemoryVectorStore();
 *
 * // Example embeddings and contexts
 * const embeddings = [
 *   [0.1, 0.2, 0.3],
 *   [0.4, 0.5, 0.6],
 *   [0.7, 0.8, 0.9]
 * ];
 * const contexts = [
 *   new Context({ pageContent: "First context", metadata: { topic: "Science" }}),
 *   new Context({ pageContent: "Second context", metadata: { topic: "Technology" }}),
 *   new Context({ pageContent: "Third context", metadata: { topic: "Mathematics" }})
 * ];
 *
 * // Add vectors to the store
 * await vectorStore.addVectors(embeddings, contexts);
 *
 * // Perform a similarity search
 * const query = [0.1, 0.1, 0.1];
 * const topK = 2;
 * const searchResults = await vectorStore.similaritySearch(query, topK);
 *
 * // Log the search results
 * console.log("Search Results:", searchResults.map(r => ({ content: r[0].pageContent, score: r[1] })));
 * ```
 */
export class MemoryVectorStore
  extends BaseVectorStore
  implements MemoryVectorStoreField
{
  /**
   * Defines the filter type for the MemoryVectorStore. Filters can include a function to
   * determine inclusion based on context, and an option to include embeddings in the returned data.
   */
  declare FilterType: MemoryVectorStoreFilterType;

  _isSerializable = true;

  /**
   * A list of all vectors currently stored in the memory. Each vector includes its content,
   * embedding, and metadata.
   */
  vectors: Vector[];

  /**
   * The function used to calculate the similarity between two vectors. This is typically a
   * cosine similarity function from a machine learning library.
   */
  similarity: typeof mlDistanceSimilarity.cosine;

  /**
   * Optional filters that can be applied during vector retrieval operations to limit the
   * results based on custom logic.
   */
  filter?: this['FilterType'];

  _vectorstoreType(): string {
    return 'memory';
  }

  constructor(fields?: Partial<MemoryVectorStoreField>) {
    super(fields);

    this.vectors = fields?.vectors ?? [];
    this.similarity = fields?.similarity ?? mlDistanceSimilarity.cosine;
    this.filter = fields?.filter;
  }

  async addVectors(
    embeddings: number[][],
    context: Context[],
    options?: { ids?: string[] }
  ): Promise<void> {
    if (embeddings.length === 0) {
      return;
    }

    if (embeddings.length !== context.length) {
      throw new Error(
        'Memory addVectors: unmatched # of embeddings and # of contexts'
      );
    }

    let ids: string[] = options?.ids ?? [];
    if (ids.length > 0 && ids.length > embeddings.length) {
      console.warn(
        'Memory addVectors: # of ids is larger than # of embeddings. Now truncate last few ids'
      );

      ids = ids.slice(0, embeddings.length);
    } else if (ids.length > 0 && ids.length < embeddings.length) {
      console.warn(
        'Memory addVectors: # of ids is less than # of embeddings. Now pad last few ids with some random ids'
      );

      const paddingArray = Array.from(
        { length: embeddings.length - ids.length },
        () => getRecordId()
      );
      ids = ids.concat(paddingArray);
    } else if (ids.length === 0) {
      ids = Array.from({ length: embeddings.length }, () => getRecordId());
    }

    const newVectors: Vector[] = embeddings.map((embedding, index) => ({
      id: ids[index],
      content: context[index].pageContent,
      embedding,
      metadata: context[index].metadata,
    }));

    this.vectors = this.vectors.concat(newVectors);
  }

  async deleteVectors(
    params?: MemoryDeleteParams<this['FilterType']>
  ): Promise<void> {
    if (params?.ids || params?.filter) {
      let filteredVectors: Vector[] = this.vectors;

      if (params?.filter) {
        const _filter = params?.filter;

        const filterFunc = async (vector: Vector) => {
          if (!_filter.filterFunc) {
            return true;
          }

          const context = new Context({
            pageContent: vector.content,
            metadata: vector.metadata,
          });

          return _filter.filterFunc(context);
        };

        filteredVectors = filteredVectors.filter(filterFunc);
      }

      if (params?.ids) {
        const _ids: string[] = params?.ids ?? [];

        filteredVectors = filteredVectors.filter((v) => _ids.includes(v.id));
      }

      const vectorIdsToDelete: string[] = filteredVectors.map((v) => v.id);

      this.vectors = this.vectors.filter(
        (v) => !vectorIdsToDelete.includes(v.id)
      );
    } else {
      throw new Error(
        "Chroma deleteVectors: must provide either 'ids' or 'filter'"
      );
    }
  }

  async similaritySearch(
    query: number[],
    topK: number,
    filter?: this['FilterType']
  ): Promise<[Context, number][]> {
    const _filter = filter ?? this.filter;

    const filterFunc = async (vector: Vector) => {
      if (!_filter?.filterFunc) {
        return true;
      }

      const context = new Context({
        pageContent: vector.content,
        metadata: vector.metadata,
      });

      return _filter.filterFunc(context);
    };

    const filteredVectors = this.vectors.filter(filterFunc);

    const searches = filteredVectors
      .map((vector, index) => ({
        similarity: this.similarity(query, vector.embedding),
        index,
      }))
      .sort((a, b) => (a.similarity > b.similarity ? -1 : 0))
      .slice(0, topK);

    return searches.map((search) => {
      const context = new Context({
        pageContent: filteredVectors[search.index].content,
        metadata: filteredVectors[search.index].metadata,
      });

      if (_filter?.includeEmbeddings) {
        context.metadata.embedding = filteredVectors[search.index].embedding;
      }

      const score = search.similarity;

      return [context, score];
    });
  }

  async maxMarginalRelevanceSearch(
    query: number[],
    topK: number,
    lambda: number,
    filter?: this['FilterType'] | undefined
  ): Promise<[Context, number][]> {
    const _filter = filter ?? this.filter;
    const includeEmbeddingsFlag = _filter?.includeEmbeddings || false;

    const searches = await this.similaritySearch(query, topK, {
      ..._filter,
      includeEmbeddings: true,
    });

    const embeddings = searches.map(
      (s) => s[0].metadata.embedding
    ) as number[][];

    const [mmrIndexes, mmrSimilarities] = maximalMarginalRelevance(
      query,
      embeddings,
      lambda,
      topK
    );

    return mmrIndexes.map((index) => {
      const context = searches[index][0];
      const similarity = mmrSimilarities[index];

      if (!includeEmbeddingsFlag) {
        delete context.metadata.embedding;
      }

      return [context, similarity];
    });
  }
}
