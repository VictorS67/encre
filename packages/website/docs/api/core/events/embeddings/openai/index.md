# events/embeddings/openai

## Index

### Classes

- [OpenAIEmbeddings](classes/OpenAIEmbeddings.md)

### Interfaces

- [OpenAIEmbeddingsCallOptions](interfaces/OpenAIEmbeddingsCallOptions.md)
- [OpenAIEmbeddingsParams](interfaces/OpenAIEmbeddingsParams.md)
