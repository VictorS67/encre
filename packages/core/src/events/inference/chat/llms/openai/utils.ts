import { OpenAI as OpenAIClient } from 'openai';

import {
  type BaseMessage,
  type ContentLike,
  MessageRole,
  ChatMessage,
} from '../../../../input/load/msgs/index.js';

/**
 * Determines if a message requires multi-modal handling based on its content. Multi-modal
 * models do not support function/tool calls.
 *
 * @param message - The message to check for modality requirements.
 * @returns A boolean indicating if multi-modal processing is needed for the message.
 * @example
 * ```typescript
 * const message = new HumanMessage({ content: "Hello" });
 * const isMultiModalNeeded = isModalityRequiredInMessage(message);
 * console.log(isMultiModalNeeded); // Output: false
 * ```
 */
export function isModalityRequiredInMessage(message: BaseMessage): boolean {
  // Multi-modal model does not suport any function/tools
  if (message._role() === 'function') {
    return false;
  }

  const isModalityRequired = (content: ContentLike): boolean => {
    return typeof content !== 'string' && isChatCompletionContentPart(content);
  };

  if (!Array.isArray(message.content)) {
    return isModalityRequired(message.content);
  }

  return message.content.some((content: ContentLike) =>
    isModalityRequired(content)
  );
}

/**
 * Checks if a part of a message matches the expected structure of a chat completion content
 * part in OpenAI's API.
 *
 * @param partLike - The object to check.
 * @returns A boolean indicating if the object is a valid chat completion content part.
 */
export function isChatCompletionContentPart(
  partLike: any
): partLike is OpenAIClient.ChatCompletionContentPart {
  return (
    typeof partLike === 'object' &&
    (isChatCompletionContentPartImage(partLike) ||
      isChatCompletionContentPartText(partLike))
  );
}

/**
 * Determines if a content part is specifically an image, based on OpenAI's chat completion
 * content structure.
 *
 * @param partLike - The object to check.
 * @returns A boolean indicating if the part is an image content part.
 */
export function isChatCompletionContentPartImage(
  partLike: any
): partLike is OpenAIClient.ChatCompletionContentPartImage {
  return (
    typeof partLike === 'object' &&
    'type' in partLike &&
    partLike['type'] === 'image_url' &&
    'image_url' in partLike &&
    isOpenAIChatImageURL(partLike['image_url'])
  );
}

/**
 * Validates whether an object is structured as an OpenAI chat image URL, including checking
 * detail levels.
 *
 * @param imageUrlLike - The object to validate.
 * @returns A boolean indicating if the object is a valid OpenAI chat image URL.
 */
export function isOpenAIChatImageURL(
  imageUrlLike: any
): imageUrlLike is OpenAIClient.ChatCompletionContentPartImage.ImageURL {
  if (typeof imageUrlLike !== 'object') return false;

  if (
    'detail' in imageUrlLike &&
    !['auto', 'low', 'high'].includes(imageUrlLike['detail'])
  ) {
    return false;
  }

  return 'url' in imageUrlLike && typeof imageUrlLike['url'] === 'string';
}

/**
 * Verifies if a part of a message is structured as a text part, according to the OpenAI chat
 * API's expectations.
 *
 * @param partLike - The object to check.
 * @returns A boolean indicating if the part is a text content part.
 */
export function isChatCompletionContentPartText(
  partLike: any
): partLike is OpenAIClient.ChatCompletionContentPartText {
  return (
    typeof partLike === 'object' &&
    'type' in partLike &&
    partLike['type'] === 'text' &&
    'text' in partLike &&
    typeof partLike['text'] === 'string'
  );
}

/**
 * Checks if an object represents a tool call within a chat message, matching the structure
 * expected by the OpenAI API.
 *
 * @param toolLike - The object to check.
 * @returns A boolean indicating if the object is a valid tool call.
 */
export function isChatCompletionMessageToolCall(
  toolLike: any
): toolLike is OpenAIClient.ChatCompletionMessageToolCall {
  return (
    typeof toolLike === 'object' &&
    'type' in toolLike &&
    toolLike['type'] === 'function' &&
    'id' in toolLike &&
    typeof toolLike['id'] === 'string' &&
    'function' in toolLike &&
    isOpenAIFunction(toolLike['function'])
  );
}

/**
 * Validates whether an object is structured as a function call for OpenAI's chat API,
 * including argument validation.
 *
 * @param functionLike - The object to validate.
 * @returns A boolean indicating if the object is a valid function structure.
 */
export function isOpenAIFunction(
  functionLike: any
): functionLike is OpenAIClient.ChatCompletionMessageToolCall.Function {
  return (
    typeof functionLike === 'object' &&
    'name' in functionLike &&
    typeof functionLike['name'] === 'string' &&
    /**
     * The following description is the description of `arguments`:
     *
     * The arguments to call the function with, as generated by the model in JSON
     * format. Note that the model does not always generate valid JSON, and may
     * hallucinate parameters not defined by your function schema. Validate the
     * arguments in your code before calling your function.
     *
     * It seems that this parameter is containing a JSON object, even the type is string.
     * So we allow object type in this parameter, but we will convert it to string before
     * calling the model API.
     */
    'arguments' in functionLike &&
    (typeof functionLike['arguments'] === 'string' ||
      typeof functionLike['arguments'] === 'object')
  );
}

/**
 * Converts a BaseMessage into a format suitable for OpenAI's Chat API, handling different
 * roles such as system, human, and assistant.
 *
 * @param message - The message to convert.
 * @returns An ChatCompletionMessageParam object representing the converted message.
 * @example
 * ```
 * const message = new HumanMessage({ content: "Example content });
 * const contentParam = getContentFromMessage(message);
 * console.log(contentParam);
 * ```
 */
export function getContentFromMessage(
  message: BaseMessage
): OpenAIClient.Chat.ChatCompletionMessageParam {
  const role: MessageRole = message._role();

  const getContentWithRole = (
    message: BaseMessage,
    role: string
  ): OpenAIClient.Chat.ChatCompletionMessageParam => {
    if (role === 'system') {
      return {
        role,
        name: message.name,
        content: coerceMessageContentToString(message.content),
      };
    } else if (role === 'human') {
      return {
        role: 'user',
        name: message.name,
        content:
          typeof message.content === 'string'
            ? message.content
            : getPartsFromMessageContent(message.content),
      };
    } else if (role === 'assistant') {
      return {
        role,
        name: message.name,
        content: coerceMessageContentToString(message.content),
        tool_calls: getToolCallsFromMessageKwargs(message.additionalKwargs),
      };
    } else if (role === 'function') {
      return {
        role: 'tool',
        content: coerceMessageContentToString(message.content),
        tool_call_id: getToolCallIdFromMessageKwargs(message.additionalKwargs),
      };
    }

    throw new Error(`CANNOT get OpenAIChat content because of role: ${role}`);
  };

  if (role !== 'general') {
    return getContentWithRole(message, role);
  }

  const generalMessage = message as ChatMessage;
  if (
    ['human', 'assistant', 'system', 'function'].includes(generalMessage.role)
  ) {
    return getContentWithRole(generalMessage, generalMessage.role);
  }

  return getContentWithRole(generalMessage, 'human');
}

/**
 * Checks if a string content is in valid JSON format.
 *
 * @param content - The content string to check.
 * @returns A boolean indicating if the content is valid JSON.
 */
export function isJSONInContent(content: string): boolean {
  try {
    JSON.parse(content);
    return true;
  } catch (e) {
    console.warn('content is not JSON');
  }

  return false;
}

/**
 * Attempts to parse and reformat a JSON string into a more structured JSON object.
 *
 * @param content - The JSON string to parse and format.
 * @returns A JSON object if parsing is successful, or null if an error occurs.
 */
export function formatJSONInContent(
  content: string
): Record<string, unknown> | null {
  try {
    return JSON.parse(content);
  } catch (e) {
    console.warn('content is not JSON, cannot format');
  }

  return null;
}

/**
 * Attempts to reformat a JSON string for better readability or further processing.
 *
 * @param content - The JSON string to format.
 * @returns A formatted JSON string.
 */
export function formatJSONStringInContent(content: string): string {
  try {
    JSON.parse(content);
    return formatArguments(content);
  } catch (e) {
    console.warn('content is not JSON, cannot format arguments');
  }

  return content;
}

/**
 * Formats a JSON string representing arguments into a more readable and properly indented string format.
 * This function is useful when you need to display or log the structured arguments in a human-readable
 * form, especially for debugging or presentation purposes.
 *
 * The function parses the JSON string to an object, iterates over its properties, and constructs a
 * formatted string with each property on a new line and indented for clarity.
 *
 * @param args - A JSON string representing the arguments that need formatting.
 * @returns A string that represents the formatted arguments in a structured, pretty-printed format.
 * @example
 * ```typescript
 * const args = '{"name":"John Doe","age":30,"skills":["JavaScript", "TypeScript"]}';
 * const formatted = formatArguments(args);
 * console.log(formatted);
 * // Output:
 * // {
 * //   "name":"John Doe",
 * //   "age":30,
 * //   "skills":["JavaScript","TypeScript"]
 * // }
 * ```
 */
export function formatArguments(args: string): string {
  const lines: string[] = [];

  lines.push('{');
  const json = JSON.parse(args);

  const properties: string[] = [];
  for (const fieldName of Object.keys(json)) {
    properties.push(`"${fieldName}":${formatValue(json[fieldName])}`);
  }

  lines.push(properties.join(',\n'));
  lines.push('}');

  return lines.join('\n');
}

function formatValue(value: unknown): string {
  if (typeof value === 'string') {
    return `"${value}"`;
  }
  if (typeof value === 'number') {
    return `${value}`;
  }
  if (Array.isArray(value)) {
    return `[${value.map((v) => formatValue(v)).join(',')}]`;
  }
  return '""';
}

function coerceMessageContentToString(
  content: ContentLike | ContentLike[]
): string {
  const coerceSingleContentLikeToString = (content: ContentLike): string => {
    if (typeof content === 'string') {
      return content;
    }

    return JSON.stringify(content, null, 2);
  };

  if (Array.isArray(content)) {
    return content.map((c) => coerceSingleContentLikeToString(c)).join('\n\n');
  }

  return coerceSingleContentLikeToString(content);
}

function getPartsFromMessageContent(
  content: ContentLike | ContentLike[]
): Array<OpenAIClient.ChatCompletionContentPart> {
  const getPart = (
    content: ContentLike
  ): OpenAIClient.ChatCompletionContentPart => {
    if (typeof content === 'string') {
      return {
        type: 'text',
        text: content,
      };
    } else if (isChatCompletionContentPartText(content)) {
      return content;
    } else if (isChatCompletionContentPartImage(content)) {
      return content;
    }

    throw new Error('Message is not valid for OpenAIChat');
  };

  if (!Array.isArray(content)) {
    return [getPart(content)];
  }

  return content.map((content: ContentLike) => getPart(content));
}

function getToolCallsFromMessageKwargs(
  additionalKwargs: Record<string, unknown> | undefined
): Array<OpenAIClient.ChatCompletionMessageToolCall> {
  if (
    additionalKwargs === undefined ||
    !('tool_calls' in additionalKwargs) ||
    !Array.isArray(additionalKwargs['tool_calls'])
  ) {
    return [];
  }

  if (
    additionalKwargs['tool_calls'].every((tc) =>
      isChatCompletionMessageToolCall(tc)
    )
  ) {
    return additionalKwargs['tool_calls'].map((tc) => ({
      ...tc,
      function: {
        ...tc.function,
        arguments:
          typeof tc.function.arguments === 'object'
            ? JSON.stringify(tc.function.arguments, null, 2)
            : tc.function.arguments,
      },
    }));
  }

  throw new Error('Message kwarg `tool_calls` is not valid for OpenAIChat');
}

function getToolCallIdFromMessageKwargs(
  additionalKwargs: Record<string, unknown> | undefined
) {
  if (
    additionalKwargs === undefined ||
    !('tool_call_id' in additionalKwargs) ||
    !(typeof additionalKwargs['tool_call_id'] === 'string')
  ) {
    console.warn(
      'Function message does not have a `tool_call_id` in OpenAI ToolMessage'
    );

    return '';
  }

  return additionalKwargs['tool_call_id'];
}
