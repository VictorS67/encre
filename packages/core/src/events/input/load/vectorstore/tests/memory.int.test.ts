import { describe, expect, test } from '@jest/globals';
import { OpenAIEmbeddings } from '../../../../embeddings/openai.js';
import { MemoryVectorStore, Vector } from '../memory.js';

describe('MemoryVectorStore', () => {
  const OPENAI_API_KEY = 'you_should_get_this_api_from_openai';

  const openaiEmbeddings = new OpenAIEmbeddings({
    openAIApiKey: OPENAI_API_KEY,
    modelName: 'text-embedding-ada-002',
  });

  const embed = async (query: string): Promise<number[]> => {
    const embedResult = await openaiEmbeddings.invoke(query);

    return embedResult.embedding;
  };

  test('similarity search', async () => {
    const memory = new MemoryVectorStore({
      vectors: [
        {
          id: 'vec1',
          content: 'I like pizza',
          embedding: await embed('I like pizza'),
          metadata: {},
        },
        {
          id: 'vec2',
          content: "I don't like dogs",
          embedding: await embed("I don't like dogs"),
          metadata: {},
        },
        {
          id: 'vec2',
          content: 'I looks like a clown',
          embedding: await embed('I looks like a clown'),
          metadata: {},
        },
      ],
    });

    const searches = await memory.similaritySearch(
      await embed('what do I like?'),
      2
    );

    expect(searches).toMatchSnapshot();
  });

  test('mmr search', async () => {
    const memory = new MemoryVectorStore({
      vectors: [
        {
          id: 'vec1',
          content: 'I like pizza',
          embedding: await embed('I like pizza'),
          metadata: {},
        },
        {
          id: 'vec2',
          content: "I don't like dogs",
          embedding: await embed("I don't like dogs"),
          metadata: {},
        },
        {
          id: 'vec2',
          content: 'I looks like a clown',
          embedding: await embed('I looks like a clown'),
          metadata: {},
        },
      ],
    });

    const searches = await memory.maxMarginalRelevanceSearch(
      await embed('what do I like?'),
      2,
      0.5
    );

    expect(searches).toMatchSnapshot();
  });
});
