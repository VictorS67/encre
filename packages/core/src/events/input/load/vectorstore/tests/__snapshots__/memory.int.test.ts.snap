// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemoryVectorStore mmr search 1`] = `
[
  [
    Context {
      "metadata": {},
      "pageContent": "I like pizza",
    },
    0.833390357575667,
  ],
  [
    Context {
      "metadata": {},
      "pageContent": "I don't like dogs",
    },
    -0.014507155088069712,
  ],
]
`;

exports[`MemoryVectorStore similarity search 1`] = `
[
  [
    Context {
      "metadata": {},
      "pageContent": "I like pizza",
    },
    0.833390357575667,
  ],
  [
    Context {
      "metadata": {},
      "pageContent": "I don't like dogs",
    },
    0.7779635310571812,
  ],
]
`;
